iroh/src/magicsock.rs

```diff
@@ -1712,64 +1713,48 @@ enum DiscoBoxError {
     Parse(anyhow::Error),
 }
 
-/// Channel for [`MagicSock::poll_recv_relay`] to receive datagrams from relays.
-///
-/// The sender and receiver will take care of the required wakers needed for
-/// [`AsyncUdpSocket::poll_recv`].
-// TODO: This channel should possibly be implemented with concurrent-queue and atomic-waker.
-// Or maybe async-channel.
-fn relay_recv_channel() -> (RelayRecvSender, RelayRecvReceiver) {
-    let (tx, rx) = mpsc::channel(128);
-    let waker = Arc::new(parking_lot::Mutex::new(None));
-    let sender = RelayRecvSender {
-        sender: tx,
-        waker: waker.clone(),
-    };
-    let receiver = RelayRecvReceiver {
-        receiver: parking_lot::Mutex::new(rx),
-        waker,
-    };
-    (sender, receiver)
+#[derive(Debug)]
+struct RelayDatagramsQueue {
+    queue: ConcurrentQueue<RelayRecvDatagram>,
+    waker: AtomicWaker,
 }
 
-#[derive(Debug, Clone)]
-struct RelayRecvSender {
-    sender: mpsc::Sender<RelayRecvDatagram>,
-    waker: Arc<parking_lot::Mutex<Option<Waker>>>,
-}
+impl RelayDatagramsQueue {
+    fn new() -> Self {
+        Self {
+            queue: ConcurrentQueue::bounded(128),
+            waker: AtomicWaker::new(),
+        }
+    }
 
-impl RelayRecvSender {
     fn try_send(
         &self,
         item: RelayRecvDatagram,
-    ) -> Result<(), mpsc::error::TrySendError<RelayRecvDatagram>> {
-        self.sender.try_send(item).inspect(|_| {
-            if let Some(waker) = self.waker.lock().take() {
-                waker.wake();
-            }
+    ) -> Result<(), concurrent_queue::PushError<RelayRecvDatagram>> {
+        self.queue.push(item).inspect(|_| {
+            self.waker.wake();
         })
     }
-}
-
-#[derive(Debug)]
-struct RelayRecvReceiver {
-    receiver: parking_lot::Mutex<mpsc::Receiver<RelayRecvDatagram>>,
-    waker: Arc<parking_lot::Mutex<Option<Waker>>>,
-}
 
-impl RelayRecvReceiver {
     fn poll_recv(&self, cx: &mut Context) -> Poll<Result<RelayRecvDatagram>> {
-        let mut receiver = self.receiver.lock();
-        self.waker.lock().replace(cx.waker().clone());
-        match receiver.try_recv() {
-            Ok(item) => {
-                self.waker.lock().take();
-                Poll::Ready(Ok(item))
-            }
-            Err(mpsc::error::TryRecvError::Empty) => Poll::Pending,
-            Err(mpsc::error::TryRecvError::Disconnected) => {
-                Poll::Ready(Err(anyhow!("All RelayRecvSenders disconnected")))
+        match self.queue.pop() {
+            Ok(value) => Poll::Ready(Ok(value)),
+            Err(concurrent_queue::PopError::Empty) => {
+                self.waker.register(cx.waker());
+
+                match self.queue.pop() {
+                    Ok(value) => {
+                        self.waker.take();
```

Comment:
FWIW the Flag examples in `AtomicWaker` don't bother with removing this waker if it is no longer needed.  I really don't know anymore how much this matters.  It depends on what suprious wakeups might happen or how concurrently this is called I guess.

This *should* be reasonably not-concurrently called IIUC, but then I don't understand why I had things not waking up in the broken version.

Reply:
Well, yeah, not removing the waker here means you're generating another spurious wakeup AFAIU. Not terrible, but also not great.

Reply:
Of course custom poll functions could mean some other poll decides to still return pending for other reasons.  In that case you'll get spurious wakeups I guess.

Reply:
Yes it's exactly that case you describe. If another poll fn decides to return pending for other reasons, then there's no need to wake up the future if there's another item in the queue. The only reason to wake up the future is for the other poll fn's reasons.